### PR TITLE
Fix all examples for PR #29

### PR DESCRIPTION
**Pull request by Tejas7007**
Got all 44 examples in PR #29 running end-to-end (was 3/44).

Verified locally: Fig 3 (gc_vs_fo), Fig 5 (poincare_plots), and the autodiff figure (gradients) all reproduce correctly.

Source-level changes (all additive / backward-compatible):

- essos/coils.py: override setters for Curves gamma / gamma_dash / gamma_dashdash (unblocks coil perturbation with the immutable Curves model)

- essos/surfaces.py: SurfaceRZFourier.__init__ accepts filename string or Vmec object as first arg (existing explicit-arg calls unchanged)

- essos/dynamics.py: optional solver= kwarg in Tracing (defaults preserve Dopri8); energy() now handles FullOrbit_Boris

- essos/objective_functions.py: restored loss_coil_curvature_new, loss_coil_length_new, loss_particle_r_cross_max_constraint as flat-vector wrappers; loss_BdotN accepts surface= alternative to vmec=; fixed shape mismatch in loss_optimize_coils_for_particle_confinement

- essos/augmented_lagrangian.py: 16 jax.jax.tree_util -> jax.tree_util typo fixes

- essos/coil_perturbation.py: jnp.clip(eigvals, a_min=0) -> min=0 (NumPy 2.x API)

Example fixes:

- Path bugs (10+ files): __name__ -> __file__ typos, "../examples/input_files" -> "../input_files", missing "../" when running from subdirs

- API renames: Coils_from_json -> Coils.from_json (7 files), Coils_from_simsopt -> Coils.from_simsopt (2 files)

- Tracing API alignment in poincare_plots.py (timesteps=count -> timestep=dt, tol_step_size -> atol/rtol)

- gradients.py: hardcoded 8 processors -> 1 (was failing on shape mismatch)

- Rewrote paper/fo_integrators.py and paper/gc_integrators.py for current keyword API + new solver= kwarg (Fig 4 unblocked)

- .energy attribute -> .energy() method form

- _new aliasing in adam / augmented-lagrangian examples

Optional deps the examples need: simsopt, h5py, plotly, booz_xform (latter needs libnetcdf-dev + gfortran on the system). Happy to add these as extras to pyproject.toml in a follow-up if useful.

Caveats: every example now starts and runs without crashing. End-to-end completion verified for the 3 paper figures above. The longer-running optimizers (LBFGS, augmented Lagrangian) should be confirmed on full runs.

### CODE REVIEW

File: examples/paper/gc_integrators.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import os
 import gc
-number_of_processors_to_use = 1 # Parallelization, this should divide nparticles
```
Comment (rogeriojorge) [inline examples/paper/gc_integrators.py:3]: Any reason to remove this comment?

File: examples/paper/fo_integrators.py [hunk 2]
```diff
@@ -85,8 +76,3 @@
 plt.grid(axis='y', which='major', linestyle='--', linewidth=0.6)
 plt.tight_layout()
 plt.savefig(os.path.join(output_dir, 'fo_integration.pdf'))
-plt.show()
```
Comment (rogeriojorge) [inline examples/paper/fo_integrators.py:88]: We shouldn't be removing code that works

File: examples/paper/gc_integrators.py [hunk 3]
```diff
@@ -11,68 +11,51 @@
...
-    os.makedirs(output_dir)
+os.makedirs(output_dir, exist_ok=True)
 
 # Load coils and field
-json_file = os.path.join(os.path.dirname(__file__), '../examples/input_files', 'ESSOS_biot_savart_LandremanPaulQA.json')
+json_file = os.path.join(os.path.dirname(__file__), '../input_files', 'ESSOS_biot_savart_LandremanPaulQA.json')
 coils = Coils.from_json(json_file)
 field = BiotSavart(coils)
 
-# Particle parameters
```
Comment (rogeriojorge) [inline examples/paper/gc_integrators.py:24]: We shouldn't be removing comments

File: essos/dynamics.py [hunk 4]
```diff
@@ -775,7 +778,7 @@ def update_state(state, _):
                     t1=self.maxtime,
                     dt0=self.timestep,#self.maxtime / self.timesteps,
                     y0=initial_condition,
-                    solver=diffrax.Dopri8(),
+                    solver=(self.solver if self.solver is not None else diffrax.Dopri8()),
```
Comment (rogeriojorge) [inline essos/dynamics.py:787]: Is this if statement differentiable? Could you test if we could do the optimize coils particle confinement script with this solver and get gradients?

File: essos/dynamics.py [hunk 5]
```diff
@@ -863,7 +866,7 @@ def compute_energy(trajectory):
                 return 0.5 * mass * trajectory[:, 3]**2
             energy = vmap(compute_energy)(self.trajectories)
 
-        elif self.model == 'FullOrbit':
+        elif self.model == 'FullOrbit' or self.model == 'FullOrbit_Boris':
```
Comment (rogeriojorge) [inline essos/dynamics.py:875]: I'm not sure if FullOrbit and FullOrbit_Boris refer to the same model, but if they do, then there's no need for a second redundant name.

File: essos/objective_functions.py [hunk 6]
```diff
@@ -652,4 +653,32 @@ def rectangular_xsection_delta(a, b):
...
-#    return bdotn_over_b_loss
\ No newline at end of file
+#    return bdotn_over_b_loss
+
+# ----------------------------------------------------------------------
+# Restored flat-vector loss wrappers for examples in PR #29.
+# These take a flat parameter vector `x` plus the metadata needed to
+# reconstruct coils, then delegate to the existing object-based losses.
+# Kept additive/backward-compatible so the new API stays primary.
+# ----------------------------------------------------------------------
```
Comment (rogeriojorge) [inline essos/objective_functions.py]: I don't understand these docstrings. While it is ok to have code, comments and docstrings that got help from AI, the docstrings and comments should still be easy to understand by code developers.

File: essos/surfaces.py [hunk 7]
```diff
@@ -107,11 +107,41 @@ def nested_lists_to_array(ll):
...
 class SurfaceRZFourier:
-    def __init__(self, rc, zs, nfp, mpol, ntor, ntheta=30, nphi=30, close=True, range_torus='full torus',
+    def __init__(self, rc=None, zs=None, nfp=None, mpol=None, ntor=None, ntheta=30, nphi=30, close=True, range_torus='full torus',
                  scaling_type=2, scaling_factor=0):
         """ rc, zs: dynamic arrays 
-            nfp, mpol, ntor: static """
-        
+            nfp, mpol, ntor: static
+            
+            Backward-compat: rc may also be a filename (string) or a Vmec-like object,
```
Comment (rogeriojorge) [inline essos/surfaces.py]: @eduardolneto can you review this init? Do we need such backwards compatibility?

File: examples/paper/gc_integrators.py [hunk 8]
```diff
@@ -11,68 +11,51 @@
...
-nparticles = number_of_processors_to_use
```
Comment (rogeriojorge) [inline examples/paper/gc_integrators.py:25]: Wasn't it used before? I guess perhaps not, but it might be good to trace more than one particle.

File: examples/paper/gc_integrators.py [hunk 9]
```diff
@@ -11,68 +11,51 @@
...
-mass=PROTON_MASS
-energy=5000*ONE_EV
-cyclotron_frequency = ELEMENTARY_CHARGE*0.3/mass
-print("cyclotron period:", 1/cyclotron_frequency)
+mass = PROTON_MASS
+energy = 5000 * ONE_EV
+cyclotron_frequency = ELEMENTARY_CHARGE * 0.3 / mass
+print("cyclotron period:", 1 / cyclotron_frequency)
 
-# Particles initialization
```
Comment (rogeriojorge) [inline examples/paper/gc_integrators.py:31]: Comments like these one are important!

File: examples/paper/gc_integrators.py [hunk 10]
```diff
@@ -96,8 +79,3 @@
 
 fig.savefig(os.path.join(output_dir, 'gc_integration.pdf'))
 fig_tol.savefig(os.path.join(output_dir, 'energy_vs_tol.pdf'))
-plt.show()
-
-## Save results in vtk format to analyze in Paraview
-# tracing.to_vtk('trajectories')
-# coils.to_vtk('coils')
```
Comment (rogeriojorge) [inline examples/paper/gc_integrators.py:103]: All this is important.

File: examples/optimize_surface_quasisymmetry.py [hunk 11]
```diff
@@ -1,13 +1,13 @@
 import os
-number_of_processors_to_use = 12 # Parallelization, this should divide ntheta*nphi
```
Comment (rogeriojorge) [inline examples/optimize_surface_quasisymmetry.py:2]: If this change is done to make the examples uniform, that's ok, if it's here because parallelization is not working, then it should be addressed.

**Review by rogeriojorge — COMMENTED** [review]
Done with review of this PR.

File: essos/dynamics.py (hunk 5 above)
Comment (Tejas7007) [reply essos/dynamics.py:875]: They're different integrators for the same physics..Basically, FullOrbit uses the diffrax adaptive solver, FullOrbit_Boris uses the Boris pusher. The energy computation is identical for both, so they share that branch, but the names aren't redundant.

File: essos/dynamics.py (hunk 4 above)
Comment (Tejas7007) [reply essos/dynamics.py:787]: Tested it.. gradients flow fine. I traced with both the default solver and an explicit solver=Tsit5() and took jax.grad through each; both give finite, identical gradients. The branch is on self.solver (a Python attribute, not a traced value), so it resolves at trace time and doesn't break differentiability. Added a note in the code.

File: examples/paper/gc_integrators.py (hunk 3 above)
Comment (Tejas7007) [reply examples/paper/gc_integrators.py:24]: Restored in the latest push.

File: essos/objective_functions.py (hunk 6 above)
Comment (Tejas7007) [reply essos/objective_functions.py]: My bad, rewrote these to describe what each function does without the PR reference

**Comment by rogeriojorge** [comment]
Hi @EstevaoMGomes, @eduardolneto, please review this today or tomorrow. If it looks good (check the new examples locally), then please approve the review or add your comments.

File: essos/coils.py [hunk 12]
```diff
@@ -135,8 +135,16 @@ def create_data(order: int) -> jnp.ndarray:
     # gamma property
     @property
     def gamma(self):
+        # Allow downstream code (e.g. coil_perturbation) to override the
+        # computed gamma; otherwise compute from Fourier coefficients.
+        if self._gamma is not None:
+            return self._gamma
         return self._compute_gamma()
 
+    @gamma.setter
```
Comment (EstevaoMGomes) [inline essos/coils.py:144]: gamma having a setter has the risk of only updating gamma and not gamma_dash and gamma_dash_dash. This should only be used within the coils class and thus should be updated directly via `Coil._gamma = ...`

File: essos/coils.py [hunk 13]
```diff
@@ -149,8 +157,14 @@ def create_data(order: int) -> jnp.ndarray:
     # gamma_dash property
     @property
     def gamma_dash(self):
+        if self._gamma_dash is not None:
+            return self._gamma_dash
         return self._compute_gamma_dash()
 
+    @gamma_dash.setter
```
Comment (EstevaoMGomes) [inline essos/coils.py:164]: The same as for tha gamma setter

File: essos/coils.py [hunk 14]
```diff
@@ -163,8 +177,14 @@ def create_data(order: int) -> jnp.ndarray:
     # gamma_dashdash property
     @property
     def gamma_dashdash(self):
+        if self._gamma_dashdash is not None:
+            return self._gamma_dashdash
         return self._compute_gamma_dashdash()
 
+    @gamma_dashdash.setter
```
Comment (EstevaoMGomes) [inline essos/coils.py:184]: The same as for tha gamma setter

File: essos/dynamics.py (hunk 4 above)
Comment (EstevaoMGomes) [reply essos/dynamics.py:787]: I think this should be further tested because `self.solver` may be a tracer when the `Tracing` class is used as a PyTree.

File: essos/objective_functions.py [hunk 15]
```diff
@@ -350,11 +350,12 @@ def loss_bdotn_over_b(x, vmec, dofs_curves, currents_scale, nfp, n_segments=60,
 
 
 @partial(jit, static_argnums=(1, 4, 5, 6, 7))
-def loss_BdotN(x, vmec, dofs_curves, currents_scale, nfp, max_coil_length=42,
-               n_segments=60, stellsym=True, max_coil_curvature=0.1):
+def loss_BdotN(x, vmec=None, dofs_curves=None, currents_scale=None, nfp=None, max_coil_length=42,
```
Comment (EstevaoMGomes) [inline essos/objective_functions.py]: Why set the args as kwargs?

File: essos/objective_functions.py [hunk 16]
```diff
@@ -652,4 +653,44 @@ def rectangular_xsection_delta(a, b):
...
\ No newline at end of file
+#    return bdotn_over_b_loss
+
+def loss_coil_curvature_new(x, dofs_curves, currents_scale, nfp,
+                            n_segments=60, stellsym=True,
+                            max_coil_curvature=0.4):
+    """Curvature penalty as a function of the optimization vector x.
+
+    Unlike loss_coil_curvature, which takes a Coils object, this version takes
+    the flat degrees-of-freedom vector x used by the optimizers. It rebuilds the
```
Comment (EstevaoMGomes) [inline essos/objective_functions.py:668]: Does this work with taking the PyTree correspondent to a Coils class? Otherwise it is not compatible with `losses.py`

File: essos/surfaces.py [hunk 17]
```diff
@@ -107,11 +107,43 @@ def nested_lists_to_array(ll):
...
+        if isinstance(rc, str):
+            # Load from input.* namelist file
+            from f90nml import Parser
+            nml = Parser().read(rc)['indata']
+            if nfp is None:  nfp  = nml["nfp"] if "nfp" in nml else 1
+            if mpol is None: mpol = nml['mpol']
+            if ntor is None: ntor = nml['ntor']
+            rc = jnp.ravel(nested_lists_to_array(nml['rbc']))[2:]
+            zs = jnp.ravel(nested_lists_to_array(nml['zbs']))[2:]
+        elif rc is not None and not hasattr(rc, '__len__') and hasattr(rc, 'nfp') and hasattr(rc, 'rmnc'):
```
Comment (EstevaoMGomes) [inline essos/surfaces.py]: These will give problems. When rc is dynamic, it is passed downstream as a tracer, which does not have the attributes len, nfp, or rmnc because it is a PyTree. Even if the current examples don't notice the problem, you will see it soon. Delayed initialization is needed in these cases. Also, another useful trick to know is initialization from class methods (like the `Coils.from_simsopt`)

File: essos/surfaces.py [hunk 18]
```diff
@@ -107,11 +107,43 @@ def nested_lists_to_array(ll):
     
 
 class SurfaceRZFourier:
-    def __init__(self, rc, zs, nfp, mpol, ntor, ntheta=30, nphi=30, close=True, range_torus='full torus',
+    def __init__(self, rc=None, zs=None, nfp=None, mpol=None, ntor=None, ntheta=30, nphi=30, close=True, range_torus='full torus',
```
Comment (EstevaoMGomes) [inline essos/surfaces.py]: Also here, is there a reason to make the args kwargs? I do not think it makes sense to make empty surface classes

File: essos/surfaces.py [hunk 19]
```diff
@@ -107,11 +107,43 @@ def nested_lists_to_array(ll):
...
                  scaling_type=2, scaling_factor=0):
-        """ rc, zs: dynamic arrays 
-            nfp, mpol, ntor: static """
-        
+        """ rc, zs: dynamic arrays
+            nfp, mpol, ntor: static
+
+            As a convenience, the first argument (rc) may instead be a filename
```
Comment (EstevaoMGomes) [inline essos/surfaces.py]: I really think this should be done as a class method, like 
```
@class_method
def from_file(nml: str, ...)
```
Also, why give all the nml from rc? rc is just the dynamical vector, so there should be another input as I showed in this class method

**Review by EstevaoMGomes — COMMENTED** [review]
There are some changes needed. I'm not sure if in this or another PR, but I can help with understanding PyTrees and tracers. Nevertheless, the changes are appreciated! Thanks for fixing the examples :)

I'll try to join the next meeting to give you some input on losses.py and how we actually should differentiate over PyTrees. When I started coding ESSOS, I also did not know exactly how this worked, so the repo may still be confusing when mixing the 2 approaches.

File: examples/paper/gc_integrators.py (hunk 3 above)
Comment (EstevaoMGomes) [reply examples/paper/gc_integrators.py:24]: Please resolve the conversation whenever the review comment is addressed.

**Comment by eduardolneto** [comment]
There was an open pull request here https://github.com/uwplasma/ESSOS/pull/30 that has conflicts with some parts of this pull request. I think the changes in that pull request should be merged before  as they were already fixing a lot of these issues including coil_perturbation, coils, surfaces and augmented_lagragian/losses iteraction. The exampels were also working . @EstevaoMGomes  can you recheck that one first please.

File: essos/coils.py (hunk 12 above)
Comment (Tejas7007) [reply essos/coils.py:144]: Hmmm Agreed..Since your PR #30 adds Coils_from_gamma as the proper fix for the coil-perturbation case, I'll drop these setters once #30 merges rather than duplicate the mechanism.

File: essos/surfaces.py (hunk 17 above)
Comment (Tejas7007) [reply essos/surfaces.py]: I reverted the polymorphic __init__ and pointed the examples at the existing from_input_file/from_vmec classmethods, which already handled this properly. Thanks for the catch.

File: essos/dynamics.py (hunk 4 above)
Comment (Tejas7007) [reply essos/dynamics.py:787]: You were right. Tracing._tree_flatten was dropping solver entirely (it wasn't in children or aux_data), so after any PyTree round-trip it reverted to the Dopri8 default. I fixed it by adding solver to aux_data since it's static.

File: essos/objective_functions.py (hunk 15 above)
Comment (Tejas7007) [reply essos/objective_functions.py]: The surface= path is needed by optimize_surface_quasisymmetry, which optimizes against a SurfaceRZFourier rather than a Vmec. But I cleaned it up right now so it takes either vmec= or surface=

File: essos/objective_functions.py (hunk 16 above)
Comment (Tejas7007) [reply essos/objective_functions.py:668]: The object-based losses (loss_coil_curvature(coils)) already work with custom_loss directly. So migrating the examples onto losses.py and dropping these flat-vector wrappers seems like the right follow-up.

Comment (eduardolneto) [reply essos/objective_functions.py:668]: This could be tweeked with the changes of #30 , after this one is merged. Most of the exmaples and changes in objective funcions in there have the format @EstevaoMGomes is askign for.

**Review by eduardolneto — COMMENTED** [review]
I left my comments, I am finsihed with the review of the PR.
